---
title: "DGE by condition within oligodendroglia clusters - forebrain tissues combined"
author: "Luise A. Seeker/Sunniva Bostrand"
date: "05/11/2021"
output:
  html_document:
    toc: true
    toc_float:
      collapsed: false
    toc_depth: 4
    theme: united
---

# Load libraries
```{r}
library(Seurat)
library(here)
library(dplyr)
```

```{r}
Oligos <- readRDS("/exports/eddie/scratch/s1359339/HD/06_swapped/subclusters/20211005_Oligos_qc_integrated_annot_swappedsamples.rds")
Oligos <-Oligos[,Oligos$Tissue != "CB"]
seur_comb <- readRDS("/exports/eddie/scratch/s1359339/HD/06_swapped/20211005_seur_comb_qc_integrated_annot_swappedsamples.rds")
seur_comb$clusters_named <- Idents(seur_comb)
```


# Perform differential gene expression without splitting by tissue
```{r}
# create folder where to save data
dir.create(here("outs"))
dir.create(here("outs", "DGE_condition"))
dir.create(here("outs", "DGE_condition","within_cluster"))
dir.create(here("outs", "DGE_condition", "within_cluster", "condition"))
Idents(Oligos) <- "ol_clusters_named"
cluster_lev <- levels(Oligos@meta.data$ol_clusters_named)
for(i in 1 : length(cluster_lev)){
  clu_dat <- subset(Oligos, ident = cluster_lev[i])
  Idents(clu_dat) <- "donor_status"
  mark_condition <- FindAllMarkers(clu_dat, 
                         only.pos = F, 
                         min.pct = 0.25, 
                         logfc.threshold = 0.25, 
                         test.use = "MAST")
  save_name_condition <- paste(cluster_lev[i], "condition_mark.csv", sep = "_")
  write.csv(mark_condition, here("outs", "DGE_condition", "within_cluster", 
                               "condition", save_name_condition))
}
```


#Read in data for plotting differentially expressed genes
```{r}
# define function
gen_mark_list <-function(file_dir = getwd(),
                         avg_log = 1.2,
                         pct_1 = 0.25,
                         pct_2 = 0.6,
                         p_adj = 0.05,
                         pairwise = FALSE
){
  temp = list.files(path = file_dir,
                    pattern="*.csv")
  myfiles = lapply(paste(file_dir, temp, sep = "/"), read.csv)
  
  for(i in 1:length(myfiles)){
    dat <- myfiles[[i]]
    av_log_fil <- subset(dat, abs(dat$avg_log2FC) > avg_log & 
                           dat$pct.1 > pct_1 & 
                           dat$pct.2 < pct_2 & 
                           dat$p_val_adj < p_adj) 
    if(pairwise == TRUE){
      top10 <- av_log_fil %>% top_n(10, avg_log2FC)
      top10$gene <- top10$X
      ol_clu <- strsplit(temp[i], "_condition_mark.")[[1]][1]
      top10$ol_clu <- ol_clu
    }else{
      av_log_fil$cluster <- as.character(av_log_fil$cluster)
      fil <- av_log_fil %>% group_by(cluster)
      ol_clu <- strsplit(temp[i], "_condition_mark.")[[1]][1]
      fil$ol_clu <- ol_clu
    }
    
    if(i == 1){
      fil_genes <- fil
    }else{
      fil_genes <- rbind(fil_genes, fil)
    }
    
    fil_genes <- fil_genes[fil_genes$cluster == "HD",]
    
  }
  
  return(fil_genes)
}

# use function
fil_genes_condition <- gen_mark_list(file_dir = here("outs", "DGE_condition", "within_cluster", "condition"),
                           avg_log = 0.8,
                           pct_1 = 0.25,
                           pct_2 = 0.6,
                           p_adj = 0.05)
fil_genes <-  fil_genes_condition

int_genes <- unique(fil_genes$gene)
write.csv(fil_genes, file = "/exports/eddie/scratch/s1359339/HD/DEgenes_oligos_alltissues.csv")
```


```{r}
sessionInfo()
```
