---
title: "Look at spliced/unspliced transcripts for PDE1A"
author: "Luise A. Seeker/Sunniva Bostrand"
date: "10/06/2022"
output: html_document
---

# Introduction

```{r cars}
setwd("/exports/eddie/scratch/s1359339/HD/06_swapped")
library(Seurat)
library(dplyr)
library(here)
library(ggsci)
library(scater)
library(RColorBrewer)
```
```{r}
mypal <- pal_npg("nrc", alpha = 0.7)(10)
mypal2 <-pal_tron("legacy", alpha = 0.7)(7)
mypal3 <- pal_lancet("lanonc", alpha = 0.7)(9)
mypal4 <- pal_simpsons(palette = c("springfield"), alpha = 0.7)(16)
mypal5 <- pal_rickandmorty(palette = c("schwifty"), alpha = 0.7)(6)
mypal6 <- pal_futurama(palette = c("planetexpress"), alpha = 0.7)(5)
mypal7 <- pal_startrek(palette = c("uniform"), alpha = 0.7)(5)

mycoloursP<- c(mypal, mypal2, mypal3, mypal4, mypal5, mypal6, mypal7, "black", "blue")
```

```{r}
spliced <- readRDS(here("02_scater_cell_filtered/spliced/20210122_SCE_spliced_gene_cell_filtered_less_strict.RDS"))
unspliced <- readRDS(here("02_scater_cell_filtered/unspliced/2021022_SCE_unspliced_gene_cell_filtered_less_strict.RDS"))


oligos <- readRDS(here("subclusters",
                    "20211005_Oligos_qc_integrated_annot_swappedsamples.rds"))
oligos <- RenameIdents(oligos,   "Oligo_cedar" = "Oligo_Cedar",
                                                       "Oligo_rowan" = "Oligo_Rowan",
                                                        "Oligo_ash" = "Oligo_Ash",
                                                        "Oligo_birch" = "Oligo_Birch",
                                                        "Oligo_maple" = "Oligo_Maple", 
                                                        "Oligo_elder" = "Oligo_Elder",
                                                        "Oligo_oak" = "Oligo_Oak",
                                                       "OPC_pine" = "OPC_Pine",
                                                        "OPC_fir" = "OPC_Fir",
                                                        "COP_cherry" = "COP_Cherry",
                                                        "COP_apple" = "COP_Apple")
oligos$ol_clusters_named <- Idents(oligos)
DefaultAssay(oligos) <- "RNA"

pos<- grep("PDE1A", rownames(unspliced))
pos

pos_2<- grep("PDE1A", rownames(spliced))
pos_2

pos == pos_2
```




```{r}
unspliced_PDE1A <- unspliced[,which(assay(unspliced)['PDE1A',]  >0 )]
spliced_PDE1A <- spliced[,which(assay(spliced)['PDE1A',]  >0 )]
unspliced_PDE1A
spliced_PDE1A

```


#Try to run this in a loop to separate by disease condition
```{r}
list <- levels(as.factor(oligos$donor_status))
for (i in 1:length(list)){
  srt <- subset(oligos, donor_status == list[i])
ol_cells <- colnames(srt)
head(ol_cells)
ol_cells <- gsub('FrCx_', '', ol_cells)
ol_cells <- gsub('CB_', '', ol_cells)
ol_cells <- gsub('HC_', '', ol_cells)
ol_cells <- gsub('CN_', '', ol_cells)

head(ol_cells)

bool_1 <- colnames(unspliced_PDE1A) %in% ol_cells
ol_unspliced <- unspliced_PDE1A[,bool_1]
print(ol_unspliced)

bool_2 <- colnames(spliced_PDE1A) %in% ol_cells
ol_spliced <- spliced_PDE1A[,bool_2]
print(ol_spliced)

test_2 <- colnames(ol_spliced) %in% colnames(ol_unspliced)
summary(test_2)


df <- data.frame(mode = c("spliced", "unspliced"), 
                 counts = c(ncol(ol_spliced), ncol(ol_unspliced)), 
                 Gene = "PDE1A")
df

p <-  ggplot(df, aes(x = Gene, y = counts, fill = mode)) +
  geom_bar(position = "fill", stat = "identity") +
  theme_classic() + ylab("Proportion of total counts") + xlab("Gene") + 
  scale_fill_manual(values =mycoloursP[56:55]) + theme(axis.text = element_text(size = 20))  +
  theme(axis.title = element_text(size = 20))  + theme(legend.text = element_text(size = 15))  + theme(legend.title = element_text(size = 15)) 



name <- paste(list[i],"PDE1A_splicedunspliced.pdf", sep = "_")
            pdf(here("PhD","Thesis","thesisplots", name), width = 4.5, height = 5)
            print(p)
dev.off()

}
```



#Reads overall
```{r}

unspliced_PDE1A_only <- unspliced[which(rownames(unspliced) == "PDE1A"), ]
rowSums(assay(unspliced_PDE1A_only))


spliced_PDE1A_only <- spliced[which(rownames(spliced) == "PDE1A"),]
rowSums(assay(spliced_PDE1A_only))

(rowSums(assay(spliced_PDE1A_only))/rowSums(assay(unspliced_PDE1A_only)))*100
```

#Reads in oligos
```{r}

unspliced_PDE1A_only_ol <- ol_unspliced[which(rownames(ol_unspliced) == "PDE1A"), ]
rowSums(assay(unspliced_PDE1A_only_ol))


spliced_PDE1A_only_ol <- ol_spliced[which(rownames(ol_spliced) == "PDE1A"),]
rowSums(assay(spliced_PDE1A_only_ol))


(rowSums(assay(spliced_PDE1A_only_ol))/rowSums(assay(unspliced_PDE1A_only_ol)))*100


ratio_df_spl <- data.frame(genes = rownames(ol_spliced), 
                           gene_count = rowSums(assay(ol_spliced)),
                           modus= "spliced")

ratio_df_unspl <- data.frame(genes = rownames(ol_unspliced), 
                           gene_count = rowSums(assay(ol_unspliced)),
                           modus = "unspliced")

ratio_df <- rbind(ratio_df_spl, ratio_df_unspl)

ggplot(ratio_df, aes(x= genes, y = gene_count, fill = modus)) + 
  geom_bar(position="fill", stat="identity") +
  theme_minimal()

ggplot(ratio_df, aes(x= genes, y = gene_count, fill = modus)) + 
  geom_bar(position="stack", stat="identity") +
  theme_minimal()




all_genes_ol <-as.data.frame(rowSums(assay(ol_spliced))/rowSums(assay(ol_unspliced))*100)
names(all_genes_ol) <- "percent_spli_unspli"

all_genes_ol <- subset(all_genes_ol, all_genes_ol$percent_spli_unspli != Inf)
```

```{r}
SessionInfo()

```